modules/packages/ParallelIO.chpl

```diff
@@ -0,0 +1,705 @@
+            r.advanceTo(delim); // advance to the next delimiter
+        } catch {
+          throw new OffsetNotFoundError(
+            "Failed to find byte offset in the range (" + bounds.low:string + ".." + bounds.high:string +
+            ") for chunk " + i:string + " of " + n:string + ". Try using fewer tasks."
+          );
+        }
+
+        // try to deserialize a 't' at the offset (and the next few offsets)
+        label deserializing for 1..maxDeserTrials {
+          r.mark();
+          try {
+            const didRead = r.withDeserializer(deserializerType).read(tTest);
+            if !didRead then throw new Error();
+          } catch {
+            // couldn't deserialize a 't' at the offset
+            r.revert();   // move back to the marked position
+            r.readByte(); // advance to the next byte
+            continue deserializing; // try again
+          }
+
+          // deserialized successfully
+          r.revert();
+          chunkOffsets[i] = r.offset(); // record the starting offset
+          break advancing;
+        }
+        continue advancing;
+      }
+    }
+
+    return chunkOffsets;
+  }
+
+  /*
+    Get a prefix sum of the number of items in each chunk of the file ``f``,
+    where the chunks are defined by the ``byteOffsets`` array, and each item
+    is separated by the given delimiter.
+
+    :arg f: the file to search
+    :arg delim: the delimiter used to separate items in the file
+    :arg byteOffsets: an array of byte offsets that divide the file into chunks
+
+    :returns: an array of length ``byteOffsets.size`` containing the number of
+              items in the file before the start of each chunk. The last entry
+              contains the total number of items in the file.
+  */
+  proc findItemOffsets(const ref f: file, in delim: ?dt, const ref byteOffsets: [?d] int): [d] int throws
+    where dt == bytes || dt == string
+  {
+    var nPerChunk: [d] int;
+
+    coforall c in 0..<d.high with (ref nPerChunk) {
+      const r = f.reader(locking=false, region=byteOffsets[c]..byteOffsets[c+1]);
+
+      // count the number of items in the chunk
+      var n = 0;
+      while true {
+        try {
+          r.advanceThrough(delim); // throws if we advance to the end of the region
+          n += 1;
+        } catch {
+          break;
+        }
+      }
+      nPerChunk[c+1] = n; // (c+1) because we want an exclusive prefix-sum below
+    }
+
+    return (+ scan nPerChunk); // compute the number of items leading up to each chunk
+  }
+
+  /*
+    An error thrown when a starting offset cannot be found in a chunk of a file.
+  */
+  class OffsetNotFoundError: Error {
+    proc init() do super.init();
+    proc init(msg: string) do super.init(msg);
+  }
+
+  @chpldoc.nodoc
+  enum headerKind {
+    find,
+    nbytes,
+    nlines,
+    noHeader
+  }
+
+  /*
+    A type describing how to handle the file header when reading a file
+    in parallel.
+  */
+  record headerPolicy {
```

Comment:
I'm having a bit of trouble seeing why this should be written this way. My first intuition is that your routines should have a `headerSize` formal that allows skipping some bytes at the start of the file. I'm not sure if you really need `findStart` as well or if it can be handled with `headerSize` (say, by decrementing it by one in some cases).

If that's not workable, I think it'd be better for the functions to work with a generic type instead of this concrete type with an `enum`. I.e. `skipLines` would be a `record` and `findStart` would be a different record.

In some ways I think this is similar to the concept of

> provide the user with a way to poke around near the candidate separator to decide if it's acceptable.

that I was proposing earlier. We can consider having the user provide a record that has methods to support both uses (so there would be just one formal to customize it). Or maybe we are better of having separate records in the interface.

Reply:
I think it's nice to have support for a few different kinds of headers, and IMO this is a nice pattern for expressing those (although the implementation looks somewhat ugly):

```chapel
readParallelDelimited(f, t=T, header=headerPolicy.skipLines(1));
readParallelDelimited(f, t=T, header=headerPolicy.skipBytes(123), deserializerType=binaryDeserializer);
readParallel(f, t=T, delim="abc", header=headerPolicy.find);
readParallel(f, t=T, delim="abc", header=headerPolicy.noHeader);
```

Ideally, I was thinking it would be nice to implement this type of thing as an algebraic enum, which we don't support yet. In Rust, it would look like:

```rust
enum headerPolicy {
    skipLines(usize),
    skipBytes(usize),
    find,
    noHeader,
}
```

And "selecting" over it, looks like:
```rust
match header {
    headerPolicy::skipLines(numLines) => print!("skipping {} header lines", numLines),
    headerPolicy::skipBytes(numBytes) => print!("skipping {} header bytes", numBytes),
    headerPolicy::find => print!("finding first element"),
    headerPolicy::noHeader => print!("no header"),
}
```

Reply:
Yes my biggest concern with it is that it has no ability to be extended. A user with a different need cannot create their own way to process the header.

It is possible to have the code you showed above (e.g. `readParallel(f, t=T, delim="abc", header=headerPolicy.find);`) literally work with a different approach that uses a different record for each of these policies. Whether or not that is a good idea is a different matter.

Reply:
I see. As of now I'm not seeing exactly how that would integrate with the `findDelimChunks` and `findDelimChunksChecked` procedures. Is this a change you'd like to see before this gets merged, or something that could be modified later?

Reply:
It's OK with me if you include it as Future Work.